flink/flink.sh

```diff
@@ -22,80 +22,75 @@
 # Flink config entry: taskmanager.heap.mb
-FLINK_TASKMANAGER_MEMORY_FRACTION='0.8'
+readonly FLINK_TASKMANAGER_MEMORY_FRACTION='0.8'
+
+# Set this to true to start a flink yarn session at initialization time.
+readonly START_FLINK_YARN_SESSION="true"
+
+function configure_flink() {
+# Number of worker nodes in your cluster
+local num_workers=$(/usr/share/google/get_metadata_value attributes/dataproc-worker-count)
 
 # Determine the number of task slots
-FLINK_TASKMANAGER_SLOTS=`grep -c processor /proc/cpuinfo`
+local flink_taskmanager_slots=`grep -c processor /proc/cpuinfo`
 
 # Determine the default parallelism
-FLINK_PARALLELISM=$(python -c \
-    "print ${NUM_WORKERS} * ${FLINK_TASKMANAGER_SLOTS}")
+local flink_parallelism=$(python -c \
+    "print ${num_workers} * ${flink_taskmanager_slots}")
 
 # Calculate the memory allocations, MB, using 'free -m'. Floor to nearest MB.
-TOTAL_MEM=$(free -m | awk '/^Mem:/{print $2}')
-FLINK_JOBMANAGER_MEMORY=$(python -c \
-    "print int(${TOTAL_MEM} * ${FLINK_JOBMANAGER_MEMORY_FRACTION})")
-FLINK_TASKMANAGER_MEMORY=$(python -c \
-    "print int(${TOTAL_MEM} * ${FLINK_TASKMANAGER_MEMORY_FRACTION})")
-
-# Install Flink and tidy things up
-FLINK_TARBALL_NAME="flink-${FLINK_VERSION}-bin-hadoop${CONCAT_HADOOP_VERSION}-\
-	scala_${SCALA_VERSION}.tgz"
-FLINK_EXTRACT_DIRECTORY="flink-${FLINK_VERSION}"
-mkdir ${FLINK_INSTALL_DIR}
-cd ${FLINK_INSTALL_DIR}
-wget ${FLINK_TARBALL_URI}
-tar -zxvf ${FLINK_TARBALL_NAME}
-mv ${FLINK_EXTRACT_DIRECTORY}/* .
-rm ${FLINK_TARBALL_NAME}
-rmdir ${FLINK_EXTRACT_DIRECTORY}
+# NOTE: This should really be the memory per worker (rather than per master).
+local total_mem=$(free -m | awk '/^Mem:/{print $2}')
+local flink_jobmanager_memory=$(python -c \
+    "print int(${total_mem} * ${FLINK_JOBMANAGER_MEMORY_FRACTION})")
+local flink_taskmanager_memory=$(python -c \
+    "print int(${total_mem} * ${FLINK_TASKMANAGER_MEMORY_FRACTION})")
 
 # Apply Flink settings by appending them to the default config
 cat << EOF >> ${FLINK_INSTALL_DIR}/conf/flink-conf.yaml
 # Settings applied by Cloud Dataproc initialization action
 jobmanager.rpc.address: ${MASTER_HOSTNAME}
-jobmanager.heap.mb: ${FLINK_JOBMANAGER_MEMORY}
-taskmanager.heap.mb: ${FLINK_TASKMANAGER_MEMORY}
-taskmanager.numberOfTaskSlots: ${FLINK_TASKMANAGER_SLOTS}
-parallelism.default: ${FLINK_PARALLELISM}
+jobmanager.heap.mb: ${flink_jobmanager_memory}
+taskmanager.heap.mb: ${flink_taskmanager_memory}
+taskmanager.numberOfTaskSlots: ${flink_taskmanager_slots}
+parallelism.default: ${flink_parallelism}
 taskmanager.network.numberOfBuffers: ${FLINK_NETWORK_NUM_BUFFERS}
 fs.hdfs.hadoopconf: ${HADOOP_CONF_DIR}
 EOF
 
-# Start a Flink YARN session
-HADOOP_CONF_DIR=/etc/hadoop/conf ./bin/yarn-session.sh -n \
-	${FLINK_TASKMANAGER_SLOTS} --detached
+if [ "$START_FLINK_YARN_SESSION" = "true" ] ; then
+  env HADOOP_CONF_DIR="$HADOOP_CONF_DIR" \
+    "$FLINK_INSTALL_DIR/bin/yarn-session.sh" \
+    -n "${num_workers}" \
```

Comment:
The number of YARN containers requested was previously set to the number of CPU cores on the master (more precisely, on any node on which the init action was run). It seems like setting this to the number of worker nodes is more appropriate since we're consuming almost all of the machine's memory for each TaskManager.